### DISCUSSION TITLE
Renovate auto merges even with failing PR status checks

### DISCUSSION THREAD

**Discussion by donmahallem** [renovatebot/renovate#10366] [Request Help]
**How are you running Renovate?**

- [x] WhiteSource Renovate hosted app on github.com

**Describe the bug**

Renovate auto merges branches that have failing tests which brakes base branches as a consequence. 
One Github CI job collects the lerna packages and creates a matrix from those. Some jobs do fail within that matrix which causes the other jobs to be canceled and the finish job to be skipped. requiredStatusChecks is **not** nulled in the config
![Testmatrix](https://user-images.githubusercontent.com/4698322/121321096-17be3a80-c90e-11eb-9b1d-c0ed42b8630e.PNG)

The offending PR is: https://github.com/manniwatch/manniwatch/pull/1940

**Relevant debug logs**

<details><summary>Click me to see logs</summary>

```
{
  "state": "pending",
  "statuses": []
}
DEBUG: check runs result(branch="renovate/typescript-4.x")
{
  "checkRuns": [
    {
      "name": "Summary",
      "status": "completed",
      "conclusion": "success"
    },
    {
      "name": "Collect Lerna Packages",
      "status": "completed",
      "conclusion": "success"
    },
    {
      "name": "Label Package PRs",
      "status": "completed",
      "conclusion": "success"
    },
    {
      "name": "Label Package PRs",
      "status": "completed",
      "conclusion": "success"
    },
    {
      "name": "Collect Lerna Packages",
      "status": "completed",
      "conclusion": "success"
    }
  ]
}
DEBUG: Processing existing PR(branch="renovate/typescript-4.x")
DEBUG: getBranchStatus(renovate/typescript-4.x)(branch="renovate/typescript-4.x")
DEBUG: branch status check result(branch="renovate/typescript-4.x")
{
  "state": "pending",
  "statuses": []
}
DEBUG: check runs result(branch="renovate/typescript-4.x")
{
  "checkRuns": [
    {
      "name": "Summary",
      "status": "completed",
      "conclusion": "success"
    },
    {
      "name": "Collect Lerna Packages",
      "status": "completed",
      "conclusion": "success"
    },
    {
      "name": "Label Package PRs",
      "status": "completed",
      "conclusion": "success"
    },
    {
      "name": "Label Package PRs",
      "status": "completed",
      "conclusion": "success"
    },
    {
      "name": "Collect Lerna Packages",
      "status": "completed",
      "conclusion": "success"
    }
  ]
}
DEBUG: getBranchStatus() result(branch="renovate/typescript-4.x")
{
  "branchStatus": "green",
  "branchName": "renovate/typescript-4.x"
}
DEBUG: Pull Request #1940 does not need updating(branch="renovate/typescript-4.x")
DEBUG: PR is configured for automerge(branch="renovate/typescript-4.x")
DEBUG: getBranchStatus(renovate/typescript-4.x)(branch="renovate/typescript-4.x")
DEBUG: branch status check result(branch="renovate/typescript-4.x")
{
  "state": "pending",
  "statuses": []
}
DEBUG: check runs result(branch="renovate/typescript-4.x")
{
  "checkRuns": [
    {
      "name": "Summary",
      "status": "completed",
      "conclusion": "success"
    },
    {
      "name": "Collect Lerna Packages",
      "status": "completed",
      "conclusion": "success"
    },
    {
      "name": "Label Package PRs",
      "status": "completed",
      "conclusion": "success"
    },
    {
      "name": "Label Package PRs",
      "status": "completed",
      "conclusion": "success"
    },
    {
      "name": "Collect Lerna Packages",
      "status": "completed",
      "conclusion": "success"
    }
  ]
}
DEBUG: Automerging #1940(branch="renovate/typescript-4.x")
DEBUG: mergePr(1940, renovate/typescript-4.x)(branch="renovate/typescript-4.x")
DEBUG: mergePr(branch="renovate/typescript-4.x")
{
  "options": {
    "body": {
      "merge_method": "rebase"
    }
  },
  "url": "repos/manniwatch/manniwatch/pulls/1940/merge"
}
DEBUG: PR merged(branch="renovate/typescript-4.x")
{
  "pr": 1940
}
DEBUG: Deleted remote branch(branch="renovate/typescript-4.x")
{
  "branchName": "renovate/typescript-4.x"
}
DEBUG: No local branch to delete(branch="renovate/typescript-4.x")
{
  "branchName": "renovate/typescript-4.x"
}
INFO: PR automerged(branch="renovate/typescript-4.x")
{
  "pr": 1940,
  "prTitle": "chore(deps): update dependency typescript to v4.3.2"
}
DEBUG: No remote branch to delete(branch="renovate/typescript-4.x")
{
  "branchName": "renovate/typescript-4.x"
}
DEBUG: No local branch to delete(branch="renovate/typescript-4.x")
{
  "branchName": "renovate/typescript-4.x"
}
INFO: Renovating repository again after automerge result(branch="renovate/typescript-4.x")
DEBUG: No dangling containers to remove(branch="renovate/typescript-4.x")
```

</details>

**Have you created a minimal reproduction repository?**

Please read the [minimal reproductions documentation](https://github.com/renovatebot/renovate/blob/main/docs/development/minimal-reproductions.md) to learn how to make a good minimal reproduction repository.

- [ ] I have provided a minimal reproduction repository
- [x] I don't have time for that, but it happens in a public repository I have linked to
- [ ] I don't have time for that, and cannot share my private repository
- [ ] The nature of this bug means it's impossible to reproduce publicly

**Additional context**

Branch status checks are pending according to the log and the status checks report only the "Collect lerna packages" job but not the subsequent once. I am not sure if this is a race condition or something else is going on here. In other runs/workflows all are listed as expected.  I tried to debug this but couldn't get any clear reason.

It would be great if one could specify a (list of) job(s)/statuses to be required to finish on Github CI (in my case Finish Test). The requiredStatusChecks config option in renovate.json currently just breaks renovate with an "unsupported" error.

**Comment by HonkingGoose** [comment]
> It would be great if one could specify a (list of) job(s)/statuses to be required to finish on Github CI (in my case Finish Test).

If you setup branch protection rules on GitHub, you can tell GitHub what tests must pass before a PR can get merged. This is called "required status checks" in the branch protection screen.

[GitHub docs, require status checks before merging](https://docs.github.com/en/github/administering-a-repository/defining-the-mergeability-of-pull-requests/about-protected-branches#require-status-checks-before-merging)

I don't know if you have setup any branch protection rules for your main branch?

**Comment by donmahallem** [comment]
@HonkingGoose No, not on that repository. But that is  a good idea to circumvent the issue I have. Will set it up this way for now but still feel that renovate does something unexpected here. Or can someone explain why/what is happening here? Some times it doesn't merge this branch and sometimes it does. It feels like a race condition of some sort but not sure.

I was referring to [requiredStatusChecks](https://docs.renovatebot.com/configuration-options/#requiredstatuschecks) thrown [here](https://github.com/renovatebot/renovate/blob/a2a2f6844a74c1fc7f282b07488e18d6f9548d8c/lib/platform/github/index.ts#L887-L896). From the docs I expected it to do what I wanted but the code just disproves my expectations.

**Comment by rarkins** [comment]
@donmahallem If you look at the logs, it lists the status checks and their values, as reported by GitHub. Do you see any which were failing? Because I see "success" for all - did I miss any?

**Reply by rarkins** [reply]
Converted to discussion as this does not seem to be a bug and the title is misleading or a misunderstanding.

According to the logs, there were multiple passed tests at the time Renovate ran, and no pending tests. Therefore Renovate was completely correct to think it's OK to automerge.

Any discussions about `requiredStatusChecks` fall into the documentation or feature request category and should not be raised as a bug report.
